### ISSUE TITLE
Fail to create DB from MySQL by using MaterializeMySQL

### ISSUE THREAD

**Issue by liangli1213** [ClickHouse/ClickHouse#13991]
**Describe the bug**
clickhouse_test is a test MySQL DB, with only one table named 'name', when run "CREATE DATABASE clickhouse_test ENGINE = MaterializeMySQL('bdcmysqldev2', 'clickhouse_test', 'root', 'password');", below issue occur:

Table on ClickHouse side has been created but no data synchronized.

**Error message and/or stacktrace**
2020.08.24 15:18:17.371180 [ 38196 ] {} <Error> MaterializeMySQLSyncThread(): Query EXTERNAL DDL FROM MySQL(storage_monitor, storage_monitor) B wasn't finished successfully: Code: 62, e.displayText() = DB::Exception: Syntax error: failed at position 121 ('B'): B. Expected one of: RENAME DATABASE, TRUNCATE, RENAME TABLE, DROP query, RENAME DICTIONARY, DETACH, RENAME query, DROP, create query, CREATE, EXCHANGE TABLES, ALTER TABLE, alter query, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1a708040 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xff0e78d in /usr/bin/clickhouse
2. ? @ 0x18328b41 in /usr/bin/clickhouse
3. ? @ 0x176d06de in /usr/bin/clickhouse
4. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool) @ 0x176d27ef in /usr/bin/clickhouse
5. ? @ 0x1729f494 in /usr/bin/clickhouse
6. DB::MaterializeMySQLSyncThread::onEvent(DB::MaterializeMySQLSyncThread::Buffers&, std::__1::shared_ptr<DB::MySQLReplication::EventBase> const&, DB::MaterializeMetadata&) @ 0x172a5456 in /usr/bin/clickhouse
7. DB::MaterializeMySQLSyncThread::synchronization(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172a6d8f in /usr/bin/clickhouse
8. ? @ 0x172a7176 in /usr/bin/clickhouse
9. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xff3e4a7 in /usr/bin/clickhouse
10. ? @ 0xff3cae3 in /usr/bin/clickhouse
11. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
12. __clone @ 0xfdead in /usr/lib64/libc-2.17.so
 (version 20.8.1.4447 (official build))

**Additional context**
OS version: RHEL 7.6
Clickhouse build: 20.8.1.4447

**Comment by liangli1213** [comment]
@zhang2014 Would you please tell me the correct way to use MaterializeMySQL?

**Comment by BohuTANG** [comment]
You're missing a port after host in the first parameter.

Usage:
`CREATE DATABASE clickhouse_test ENGINE = MaterializeMySQL('mysqlhost:port', 'clickhouse_test', 'root', 'password')`

But the error message as above needs to be fixed too.

**Comment by liangli1213** [comment]
@BohuTANG , that should be typo, there is port parameter, and whats more, I do some more test:
1. CREATE DATABASE clickhouse_test ENGINE = MaterializeMySQL('bdcdev09:3306', 'clickhouse_test', 'root', 'password')
2. select * from name ORDER BY id; ---> It's works

SELECT *
FROM name
ORDER BY id ASC

┌─id─┬─username─┐
│  2 │ lli2     │
│  3 │ lli3     │
│  5 │ lli4     │
│  7 │ lli5     │
│  9 │ lli6     │
│ 11 │ lli7     │
│ 13 │ lli8     │
└────┴──────────┘
3. But when insert a new record in master (MySQL), in ClickHouse, select * from name ORDER BY id failed with error:

SELECT *
FROM name
ORDER BY id ASC

Received exception from server (version 20.8.1):
Code: 32. DB::Exception: Received from bdc-clickhouse:9000. DB::Exception: Attempt to read after eof.

So it seems bin-log replay failed.

**Comment by liangli1213** [comment]
@zhang2014 I'm using MySQL 5.7.21 and here is the my.cnf, is there anything wrong?
```
[root@bdcdev09 mysql]# cat my.cnf 
[mysqld]
basedir = /opt/mysql
datadir = /mysql/mysql5.7.21_db
socket = /var/lib/mysql/mysql.sock
pid-file = /var/lib/mysql/mysql.pid
log-error = /var/log/mysql/mysql.log
symbolic-links = 0
sql-mode = NO_ENGINE_SUBSTITUTION
log_timestamps = SYSTEM
server-id = 11100
auto_increment_offset = 1
auto_increment_increment = 2

gtid-mode = on
enforce-gtid-consistency = true
# master-info-repository = TABLE
# relay-log-info-repository = TABLE
# sync-master-info = 1
# slave-parallel-workers = 0

log-bin = mysql-bin
binlog-format = row
binlog_row_image = full
# log-slave-updates = true
binlog-checksum = none
# sync_binlog = 0
expire_logs_days = 5

binlog-ignore-db = information_schema
binlog-ignore-db = performance_schema
binlog-ignore-db = mysql
binlog-ignore-db = sys

# replicate-ignore-db = information_schema
# replicate-ignore-db = performance_schema
# replicate-ignore-db = mysql
# replicate-ignore-db = sys

# max_connect_errors = 30

# wait_timeout = 1800
# interactive_timeout = 1800
max_allowed_packet = 500M
# rpl_semi_sync_master_enabled = 1
# rpl_semi_sync_master_timeout = 2000
# rpl_semi_sync_slave_enabled = 1

######## Memory related setting ########
key_buffer_size = 64M
query_cache_size = 64M
tmp_table_size = 16M
innodb_buffer_pool_size = 128M
# innodb_additional_mem_pool_size = 8M
innodb_log_buffer_size = 16M
sort_buffer_size = 2M
read_buffer_size = 1M
read_rnd_buffer_size = 1M
join_buffer_size = 1M
thread_stack = 1M
binlog_cache_size = 1M
max_heap_table_size = 16M
max_connections = 150
########################################

# skip-name-resolve
# slow_query_log = 1
# long_query_time = 6
# slow_query_log_file = slow-query.log

[mysqld_safe]
log-error = /var/log/mysql/mysql.log
```
And here is the ClickHouse error log when I insert a new record in MySQL:
```
2020.08.24 22:52:04.147610 [ 79350 ] {} <Error> MaterializeMySQLSyncThread(clickhouse_test): Query EXTERNAL DDL FROM MySQL(clickhouse_test, clickhouse_test) B wasn't finished successfully: Code: 62, e.displayText() = DB::Exception: Syntax error: failed at position 121 ('B'): B. Expected one of: RENAME DATABASE, TRUNCATE, RENAME TABLE, DROP query, RENAME DICTIONARY, DETACH, RENAME query, DROP, create query, CREATE, EXCHANGE TABLES, ALTER TABLE, alter query, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1a708040 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xff0e78d in /usr/bin/clickhouse
2. ? @ 0x18328b41 in /usr/bin/clickhouse
3. ? @ 0x176d06de in /usr/bin/clickhouse
4. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool) @ 0x176d27ef in /usr/bin/clickhouse
5. ? @ 0x1729f494 in /usr/bin/clickhouse
6. DB::MaterializeMySQLSyncThread::onEvent(DB::MaterializeMySQLSyncThread::Buffers&, std::__1::shared_ptr<DB::MySQLReplication::EventBase> const&, DB::MaterializeMetadata&) @ 0x172a5456 in /usr/bin/clickhouse
7. DB::MaterializeMySQLSyncThread::synchronization(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172a6d8f in /usr/bin/clickhouse
8. ? @ 0x172a7176 in /usr/bin/clickhouse
9. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xff3e4a7 in /usr/bin/clickhouse
10. ? @ 0xff3cae3 in /usr/bin/clickhouse
11. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
12. __clone @ 0xfdead in /usr/lib64/libc-2.17.so
 (version 20.8.1.4447 (official build))
2020.08.24 22:52:04.147773 [ 79350 ] {} <Error> MaterializeMySQLSyncThread: Code: 62, e.displayText() = DB::Exception: Syntax error: failed at position 121 ('B'): B. Expected one of: RENAME DATABASE, TRUNCATE, RENAME TABLE, DROP query, RENAME DICTIONARY, DETACH, RENAME query, DROP, create query, CREATE, EXCHANGE TABLES, ALTER TABLE, alter query, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1a708040 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xff0e78d in /usr/bin/clickhouse
2. ? @ 0x18328b41 in /usr/bin/clickhouse
3. ? @ 0x176d06de in /usr/bin/clickhouse
4. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool) @ 0x176d27ef in /usr/bin/clickhouse
5. ? @ 0x1729f494 in /usr/bin/clickhouse
6. DB::MaterializeMySQLSyncThread::onEvent(DB::MaterializeMySQLSyncThread::Buffers&, std::__1::shared_ptr<DB::MySQLReplication::EventBase> const&, DB::MaterializeMetadata&) @ 0x172a5456 in /usr/bin/clickhouse
7. DB::MaterializeMySQLSyncThread::synchronization(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172a6d8f in /usr/bin/clickhouse
8. ? @ 0x172a7176 in /usr/bin/clickhouse
9. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xff3e4a7 in /usr/bin/clickhouse
10. ? @ 0xff3cae3 in /usr/bin/clickhouse
11. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
12. __clone @ 0xfdead in /usr/lib64/libc-2.17.so
 (version 20.8.1.4447 (official build))
2020.08.24 22:52:04.148177 [ 79350 ] {} <Error> MaterializeMySQLSyncThread: Code: 32, e.displayText() = DB::Exception: Attempt to read after eof, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1a708040 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xff0e78d in /usr/bin/clickhouse
2. DB::ReadBuffer::throwReadAfterEOF() @ 0xff563fd in /usr/bin/clickhouse
3. DB::ReadBuffer::ignore(unsigned long) @ 0x1461ff8c in /usr/bin/clickhouse
4. DB::MySQLReplication::TableMapEvent::parseImpl(DB::ReadBuffer&) @ 0x172b0f9a in /usr/bin/clickhouse
5. DB::MySQLReplication::MySQLFlavor::readPayloadImpl(DB::ReadBuffer&) @ 0x172b15bf in /usr/bin/clickhouse
6. DB::MySQLProtocol::IMySQLReadPacket::readPayload(DB::ReadBuffer&, unsigned char&) @ 0x172bc751 in /usr/bin/clickhouse
7. DB::MySQLProtocol::PacketEndpoint::tryReceivePacket(DB::MySQLProtocol::IMySQLReadPacket&, unsigned long) @ 0x172ca307 in /usr/bin/clickhouse
8. DB::MySQLClient::readOneBinlogEvent(unsigned long) @ 0x172c7c61 in /usr/bin/clickhouse
9. DB::MaterializeMySQLSyncThread::synchronization(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172a6d70 in /usr/bin/clickhouse
10. ? @ 0x172a7176 in /usr/bin/clickhouse
11. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xff3e4a7 in /usr/bin/clickhouse
12. ? @ 0xff3cae3 in /usr/bin/clickhouse
13. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
14. __clone @ 0xfdead in /usr/lib64/libc-2.17.so
 (version 20.8.1.4447 (official build))
```

**Comment by zhang2014** [comment]
More clickhouse log(before, after and during create MaterializeMySQL database). for this problem because we have received a query(=`B`) from mysql(possibly because the event was not read correctly.)

**Comment by BohuTANG** [comment]
And provide the 'name' table schema

**Comment by liangli1213** [comment]
@zhang2014 , I got the log snippet from clickhouse-server.err.log, that is the only 3 log from it for (before, after and during create MaterializeMySQL database), the previous one is 1 hour ago (I believe it has no relationship with the error)

name table schema:
![image](https://user-images.githubusercontent.com/69909759/91064687-2af36880-e662-11ea-8aff-1027567a9130.png)

**Comment by zhang2014** [comment]
Maybe you can change the log level to debug?

**Comment by liangli1213** [comment]
@zhang2014 , OK, let me try.

**Comment by zhang2014** [comment]
Can you get log from `clickhouse-server.log`(it contains debug, wraing, error levels)? 
BTW: For simplicity, maybe you can rename the original log file and repeat the above process?

``` sql
clickhouse:> DROP DATABASE clickhouse_test;
shell :> rename clickhouse-server.log clickhouse-server.log.old
shell :> restart clickhouse-server
clickhouse:> CREATE DATABASE clickhouse_test ENGINE = MaterializeMySQL('bdcmysqldev2', 'clickhouse_test', 'root', 'password');
shell :> cat clickhouse-server.log
```

**Comment by BohuTANG** [comment]
I can't reproduce it here.

1. Run MySQL with docker:
```
docker run -d -e MYSQL_ROOT_PASSWORD=123 mysql:5.7 mysqld --datadir=/var/lib/mysql --server-id=1 --log-bin=/var/lib/mysql/mysql-bin.log --gtid-mode=ON --enforce-gtid-consistency
```

2. create table:
```
mysql> Create Table: CREATE TABLE `name` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `username` varchar(40) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB

```

3. insert datas:
```
mysql> insert into name values(1,'lli1');
mysql> insert into name values(2,'lli2');
```



4. ClickHouse:
```
clickhouse> create database ck_test engine=MaterializeMySQL('172.17.0.2:3306', 'clickhouse_test', 'root', '123');
clickhouse> select * from ck_test.name;
+------+----------+
| id   | username |
+------+----------+
|    1 | lli1     |
|    2 | lli2     |
+------+----------+
2 rows in set (0.02 sec)
Read 2 rows, 62.00 B in 0.017296952 sec., 115 rows/sec., 3.50 KiB/sec.
```


5. Then insert into MySQL name table with:
```
mysql> insert into name values(3,'lli3');
mysql> insert into name(username) values('lli4');
```

6. Select from ClickHouse:
```
clickhouse> select * from ck_test.name;
+------+----------+
| id   | username |
+------+----------+
|    2 | lli2     |
|    1 | lli1     |
|    3 | lli3     |
|    4 | lli4     |
+------+----------+
4 rows in set (0.03 sec)
Read 4 rows, 124.00 B in 0.022868979 sec., 174 rows/sec., 5.30 KiB/sec.
```

Can you try this and check?

**Comment by BohuTANG** [comment]
> [clickhouse-server.log](https://github.com/ClickHouse/ClickHouse/files/5119091/clickhouse-server.log)
> Here is the log, and at last, I insert a new record, which trigger the error.

From your log your MySQL engine is `ENGINE=MyISAM`.
For replication the table engine must be InnoDB/TokuDB/MyRocks,  others engine not allowed here.

It's better to have a engine check here, cc @zhang2014

**Comment by liangli1213** [comment]
@BohuTANG , thanks, let me try InnoDB, by the way, I have 3 questions:
1. Is the below setting in MySQL a must:
> gtid-mode = on
> enforce-gtid-consistency = true
2. Will ClickHouse recover the replication from MySQL when MySQL service has been restart?
3. Will ClickHouse recover the replication when ClickHouse been restarted?
4. Is there any command to check if the replication is running? (like show slave status\G in MySQL)

**Comment by BohuTANG** [comment]
> @BohuTANG , thanks, let me try InnoDB, by the way, I have 3 questions:
> 
> 1. Is the below setting in MySQL a must:
> 
> > gtid-mode = on
> > enforce-gtid-consistency = true
> 
> 1. Will ClickHouse recover the replication from MySQL when MySQL service has been restart?
> 2. Will ClickHouse recover the replication when ClickHouse been restarted?
> 3. Is there any command to check if the replication is running? (like show slave status\G in MySQL)

1/2/3 are yes.
4 is no plan yet, maybe show as a perf event.

**Comment by liangli1213** [comment]
@BohuTANG , I have destroy the old table and created new table named 'name' with InnoDB as ENGINE, but I still met same issue.
There are so many place we need to pay attention, OS version, MySQL version, MySQL config, so what do you think we should try next step?

**Comment by liangli1213** [comment]
By the way, I use rpm to install clickhouse, as follows:

[root@bdc-clickhouse clickhouse-server]# rpm -aq | grep clickhouse
clickhouse-server-20.8.1.4447-2.noarch
clickhouse-client-20.8.1.4447-2.noarch
clickhouse-common-static-20.8.1.4447-2.x86_64

**Comment by liangli1213** [comment]
@BohuTANG @zhang2014 I found binlog-checksum = none cause the mentioned error, so after comment this setting, the DB and table can be created and data can be copied to ClickHouse.
But I met another issue, please see below error message, this happens when incremental data has been inserted into MySQL.

> 2020.08.26 10:20:34.643475 [ 42742 ] {} <Error> MaterializeMySQLSyncThread: Code: 62, e.displayText() = DB::Exception: Syntax error: failed at position 121 ('COMMIT'): COMMIT. Expected one of: RENAME DATABASE, TRUNCATE, RENAME TABLE, DROP query, RENAME DICTIONARY, DETACH, RENAME query, DROP, create query, CREATE, EXCHANGE TABLES, ALTER TABLE, alter query, Stack trace (when copying this message, always include the lines below):
> 
> 0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1a708040 in /usr/bin/clickhouse
> 1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xff0e78d in /usr/bin/clickhouse
> 2. ? @ 0x18328b41 in /usr/bin/clickhouse
> 3. ? @ 0x176d06de in /usr/bin/clickhouse
> 4. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool) @ 0x176d27ef in /usr/bin/clickhouse
> 5. ? @ 0x1729f494 in /usr/bin/clickhouse
> 6. DB::MaterializeMySQLSyncThread::onEvent(DB::MaterializeMySQLSyncThread::Buffers&, std::__1::shared_ptr<DB::MySQLReplication::EventBase> const&, DB::MaterializeMetadata&) @ 0x172a5456 in /usr/bin/clickhouse
> 7. DB::MaterializeMySQLSyncThread::synchronization(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172a6d8f in /usr/bin/clickhouse
> 8. ? @ 0x172a7176 in /usr/bin/clickhouse
> 9. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xff3e4a7 in /usr/bin/clickhouse
> 10. ? @ 0xff3cae3 in /usr/bin/clickhouse
> 11. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
> 12. __clone @ 0xfdead in /usr/lib64/libc-2.17.so
>  (version 20.8.1.4447 (official build))

**Comment by BohuTANG** [comment]
Yes.
We will take a look.

**Comment by BohuTANG** [comment]
`binlog_checksum=CRC32` is the default setting by MySQL, why you set it to NONE?

https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html#sysvar_binlog_checksum
"Setting this variable on the source to a value unrecognized by the replica causes the replica to set its own binlog_checksum value to NONE, and to stop replication with an error. (Bug #13553750, Bug #61096) If backward compatibility with older replicas is a concern, you may want to set the value explicitly to NONE."

**Comment by liangli1213** [comment]
@BohuTANG for some historical reason, it has been set to NONE, but, since I know the root cause, I will plan to change it back.
And also, I notice ClickHouse are unable to recover the data replication after MySQL restart (I'm waiting for a while) and get below error message:

> Received exception from server (version 20.8.1):
> Code: 1002. DB::Exception: Received from bdc-clickhouse:9000. DB::Exception: Master maybe lost.

**Comment by zhang2014** [comment]
> And also, I notice ClickHouse are unable to recover the data replication after MySQL restart (I'm waiting for a while) and get below error message:

DETACH DATABASE ...;
ATTACH DATABASE ...;

**Comment by liangli1213** [comment]
@zhang2014 thanks, and by the way, do you think gtid-mode = on is a must, since I switch it off and I can see the replication also works.

**Comment by liangli1213** [comment]
And also, is there anyway to stop the replication? like stop slave in MySQL.

**Comment by BohuTANG** [comment]
> @zhang2014 thanks, and by the way, do you think gtid-mode = on is a must, since I switch it off and I can see the replication also works.

Yes, other pr will support GTID based replication only, the postition is not a safe way and MySQL not prefered.
PR is here(still in review)
https://github.com/ClickHouse/ClickHouse/pull/13820

**Comment by liangli1213** [comment]
@BohuTANG so may I understand after #13820 has been merged, only GTID will be used for replication? so if GTID has been disabled, the replication will fail?

**Comment by BohuTANG** [comment]
When #13820 merged and using this release,
First, the metadata file version upgrade from 1 to 2, so you must drop the database and create again.
Second, if GTID disabled you will get the error:
```
Master Host: 172.17.0.6, Port: 3306, User: root, Password: 123, Replicate DB: dbtest, GTID: f4aee41e-e36f-11ea-8b37-0242ac110002:1-3 
Error: The replication sender thread cannot start in AUTO_POSITION mode: this server has GTID_MODE = OFF instead of ON.
```

**Comment by liangli1213** [comment]
@BohuTANG got it, thanks very much, and by the way, when do you think the GTID replication feature be merged into master?

**Comment by BohuTANG** [comment]
> @BohuTANG got it, thanks very much, and by the way, when do you think the GTID replication feature be merged into master?

If all goes well, maybe merged today :D

**Comment by liangli1213** [comment]
> 
> 
> > @BohuTANG got it, thanks very much, and by the way, when do you think the GTID replication feature be merged into master?
> 
> If all goes well, maybe merged today :D

Thanks very much, and when do you think the 20.8.2 testing build be available?

**Comment by l1t1** [comment]
where is the document of MaterializeMySQL feature?

**Comment by l1t1** [comment]
> It's merged into master, you can download the package here: https://clickhouse-builds.s3.yandex.net/0/cff92c8ad3dea2d96b1dfaadc76c3c10b3d701b2/clickhouse_build_check/report.html

i install the new rpms in the page, the version number is same as the rpms in https://repo.clickhouse.tech/rpm/testing/x86_64
```
#rpm -i clickhouse-common-static-20.8.1.4474-2.x86_64.rpm --force
# service clickhouse-server stop
Stop clickhouse-server service: DONE
# rpm -i --force clickhouse-server-20.8.1.4474-2.noarch.rpm
Path to data directory in /etc/clickhouse-server/config.xml: /var/lib/clickhouse/
# rpm -i --force clickhouse-client-20.8.1.4474-2.noarch.rpm
# service clickhouse-server start
Start clickhouse-server service: Path to data directory in /etc/clickhouse-server/config.xml: /var/lib/clickhouse/
DONE
#clickhouse-client
ClickHouse client version 20.8.1.4474 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.8.1 revision 54438.
```

**Comment by tavplubix** [comment]
> where is the document of MaterializeMySQL feature?

Documentation is not written yet. The feature is experimental and its behaviour can be slightly changed.
As for now there is a notes with some usage examples [here](https://translate.google.com/translate?sl=zh-CN&tl=en&u=https%3A%2F%2Fbohutang.me%2F2020%2F07%2F26%2Fclickhouse-and-friends-mysql-replication%2Fh) (machine-translated from Chinese).

> the version number is same

Yes, but builds are actually different:
the version from repo:
```
$ clickhouse local -q "select value from system.build_options where name='VERSION_GITHASH'"
d0b6ba35d19a1a0112b7e73a34334f329d6c04c3

$ git show d0b6ba35d19a1a0112b7e73a34334f329d6c04c3
commit d0b6ba35d19a1a0112b7e73a34334f329d6c04c3 (tag: v20.8.1.4447-testing)
Merge: f4bf99fadc 333ec2e496
Author: alexey-milovidov <milovidov@yandex-team.ru>
Date:   Mon Aug 24 01:14:34 2020 +0300

    Merge pull request #8367 from amosbird/scalarfix
    
    fix scalar subquery hash conflicts
```
the version from build check report:
```
$ clickhouse local -q "select value from system.build_options where name='VERSION_GITHASH'"
cff92c8ad3dea2d96b1dfaadc76c3c10b3d701b2

$ git show cff92c8ad3dea2d96b1dfaadc76c3c10b3d701b2
commit cff92c8ad3dea2d96b1dfaadc76c3c10b3d701b2
Merge: 318f14b95e e3bfd1a5b8
Author: tavplubix <avtokmakov@yandex-team.ru>
Date:   Wed Aug 26 15:02:53 2020 +0300

    Merge pull request #13820 from BohuTANG/mysql_replica_gtid_issue_4006
    
    ISSUES-4006 support MySQL GTID based replication #4006

```

The version number in master branch will be updated on next testing release (should happen automatically when all CI checks pass).

**Comment by l1t1** [comment]
what is the Chinese examples article link?

**Comment by BohuTANG** [comment]
> what is the Chinese examples article link?

[MySQL实时复制与实现](https://bohutang.me/2020/07/26/clickhouse-and-friends-mysql-replication/)
[MySQL实时复制之GTID模式](https://bohutang.me/2020/08/26/clickhouse-and-friends-mysql-gtid-replication/)
